pkg/operator/ceph/cluster/controller.go

```diff
@@ -809,3 +792,74 @@ func ClusterOwnerRef(namespace, clusterID string) metav1.OwnerReference {
 		BlockOwnerDeletion: &blockOwner,
 	}
 }
+
+// runHandler runs the given handler function on the cluster in a goroutine. If an already running handler is found, it will cancel the old handler before running the new one
+func (c *ClusterController) runHandler(crdName string, cluster *cluster, action string, handlerFn func(string, *cluster) bool, timeout, period time.Duration) {
+	logger.Debugf("starting %s handler for cluster in namespace %s", action, cluster.Namespace)
+
+	checkCancelHandler(cluster)
+
+	cluster.handlerCtx = newHandlerCtx(timeout)
+
+	// Start the handler using wait.PollImmediateInfinite in a goroutine.
+	// A timeout context is used to stop the handler after the time timeout.
+	// Using a timeout context allows the the wait channel to be closed when
+	// stopping the handler.
+	go wait.PollImmediateInfinite(period, func() (bool, error) {
+		select {
+		case <-cluster.handlerCtx.Done():
+			logger.Debugf("%s handler for cluster in namepsace %s is done", action, cluster.Namespace)
+			// If the context is finished, log any errors if needed and close the wait channel
+			if !cluster.handlerCtx.success {
+				var message string
+				if err := cluster.handlerCtx.Err(); err == context.Canceled {
+					message = fmt.Sprintf("canceled trying to %s cluster in namespace %s", action, cluster.Namespace)
+				} else {
+					message = fmt.Sprintf("giving up trying to %s cluster in namespace %s after %s", action, cluster.Namespace, timeout)
+				}
+				logger.Error(message)
+				if err := c.updateClusterStatus(cluster.Namespace, crdName, cephv1.ClusterStateError, message); err != nil {
+					logger.Errorf("failed to update cluster status in namespace %s: %+v", cluster.Namespace, err)
+				}
+			}
+			close(cluster.handlerCtx.wait)
+
+			// Returning true here to stop PollImmediateInfinite
+			return true, nil
+
+		default:
+			// Run handlerFn. If handlerFn finishes successfully, set the handlerCtx.success, close the wait channel and cancel the context.
+			logger.Debugf("running handlerFn for %s handler for cluster in namespace %s", action, cluster.Namespace)
+			if handlerFn(crdName, cluster) {
+				logger.Debugf("handlerFn for %s handler for cluster in namespace %s finished successfully", action, cluster.Namespace)
+
+				cluster.handlerCtx.success = true
+
+				close(cluster.handlerCtx.wait)
+				cluster.handlerCtx.cancel()
+				return true, nil
+			}
+		}
+		// Returning false here makes PollImmediateInfinite run the handler again after specified period
+		return false, nil
+	})
+
+	return
+}
+
+// checkCancelHandler will check for any running handlers for the given
+// cluster, and if found will cancel the handler and wait for it to finish
+// before returning
+func checkCancelHandler(cluster *cluster) {
```

Comment:
I found myself confused by the wording `checkCancelHandler`. I think this is better named as `cancelExistingOrchestration`. We can assume that this is idempotent and that if there is no existing orchestration, then the function is successful.